File: 1239.maximum-length-of-a-concatenated-string-with-unique-characters.cs
/*
 * @lc app=leetcode id=1239 lang=csharp
 *
 * [1239] Maximum Length of a Concatenated String with Unique Characters
 */

// @lc code=start
using System;
using System.Collections.Generic;

public class Solution
{
    /**
     * This is a DP problem
     * 1 <= arr.Length <= 16, small enough to do it semi-brute force
     * We can find all possible combinations inside arr which do not have overlapping characters
     * We use the dp list to keep all the possible used characters
     * dp[i] = all possible sets of unique characters if we use arr[0:i+1]
     */
    public int MaxLength(IList<string> arr)
    {
        int n = arr.Count;
        List<HashSet<char>> dp = new List<HashSet<char>> { new HashSet<char>() };

        foreach (string str in arr)
        {
            if (new HashSet<char>(str).Count == str.Length) // string is composed of unique characters
            {
                foreach (HashSet<char> ele in new List<HashSet<char>>(dp))
                {
                    if (ele.Overlaps(new HashSet<char>(str))) continue; // no overlap between element and the current string
                    dp.Add(new HashSet<char>(ele.Union(new HashSet<char>(str))));
                }
            }
        }

        return dp.Max(ele => ele.Count);
    }
}

// @lc code=end

File: 1239.maximum-length-of-a-concatenated-string-with-unique-characters.py
#
# @lc app=leetcode id=1239 lang=python3
#
# [1239] Maximum Length of a Concatenated String with Unique Characters
#

# @lc code=start
from typing import List


class Solution:
    '''
    This is a DP problem
    1 <= arr.length <= 16, small enough to do it semi-brute force
    we can find for all possible combinations inside arr which do not have overlapping characters
    we use the dp array to keep all the possible used characters
    dp[i] = all possible sets of unique characters if we use arr[:i+1]    
    '''
    def maxLength(self, arr: List[str]) -> int:
        n, dp = len(arr), [set()]
        for string in arr:
            if len(set(string)) == len(string): # string is composed of unique characters
                for ele in dp[:]:
                    if not ele & set(string): # no overlap between element and the current string
                        dp.append( ele | set(string))
        return max(len(ele) for ele in dp)
# @lc code=end

